File: examples/julia/schelling.jl
using Agents
using Random

const DEFAULT_GRID_W = 50
const DEFAULT_GRID_H = 50
const DEFAULT_DENSITY = 0.8
const DEFAULT_BALANCE = 0.5
const DEFAULT_SIMILARITY_THRESHOLD = 0.7

mutable struct SchellingProperties
	similarity_threshold::Float64
	last_swapped::Int
end

@agent struct SchellingAgent(GridAgent{2})
	group::Int
	agentid::String
end

function satisfied(agent::SchellingAgent, model)::Bool
	same_group = 0
	occupied_neighbors = 0

	for neighbor in nearby_agents(agent, model, 1)
		occupied_neighbors += 1
		if neighbor.group == agent.group
			same_group += 1
		end
	end

	occupied_neighbors == 0 && return true

	return same_group >= model.similarity_threshold * occupied_neighbors
end

function satisfied_pct(model)::Float64
	n = nagents(model)
	n == 0 && return 0.0
	return count(agent -> satisfied(agent, model), allagents(model)) / n
end

function segregation_index(model)::Float64
	total_ratio = 0.0
	count_agents = 0

	for agent in allagents(model)
		same = 0
		occupied_neighbors = 0

		for neighbor in nearby_agents(agent, model, 1)
			occupied_neighbors += 1
			if neighbor.group == agent.group
				same += 1
			end
		end

		if occupied_neighbors > 0
			total_ratio += same / occupied_neighbors
			count_agents += 1
		end
	end

	return count_agents == 0 ? 0.0 : total_ratio / count_agents
end

function schelling_model_step!(model)
	unsatisfied = Int[]
	for id in allids(model)
		agent = model[id]
		if !satisfied(agent, model)
			push!(unsatisfied, id)
		end
	end

	empties = collect(empty_positions(model))

	rng = abmrng(model)
	shuffle!(rng, unsatisfied)
	shuffle!(rng, empties)

	swapped = min(length(unsatisfied), length(empties))

	for i in 1:swapped
		move_agent!(model[unsatisfied[i]], empties[i], model)
	end

	abmproperties(model).last_swapped = swapped
	return swapped > 0
end

function initialize_schelling(;
	gridwidth::Int = DEFAULT_GRID_W,
	gridheight::Int = DEFAULT_GRID_H,
	density::Float64 = DEFAULT_DENSITY,
	balance::Float64 = DEFAULT_BALANCE,
	similarity_threshold::Float64 = DEFAULT_SIMILARITY_THRESHOLD,
	seed = nothing,
)
	gridwidth = gridwidth > 0 ? gridwidth : DEFAULT_GRID_W
	gridheight = gridheight > 0 ? gridheight : DEFAULT_GRID_H
	density = 0 <= density <= 1 ? density : DEFAULT_DENSITY
	balance = 0 <= balance <= 1 ? balance : DEFAULT_BALANCE
	similarity_threshold = 0 <= similarity_threshold <= 1 ?
						   similarity_threshold :
						   DEFAULT_SIMILARITY_THRESHOLD

	rng = isnothing(seed) ? Xoshiro() : Xoshiro(seed)

	space = GridSpaceSingle(
		(gridwidth, gridheight);
		periodic = false,
		metric = :chebyshev,
	)

	properties = SchellingProperties(similarity_threshold, 0)

	model = StandardABM(
		SchellingAgent,
		space;
		properties,
		model_step! = schelling_model_step!,
		rng,
	)

	next_type1 = 0
	next_type2 = 0
	type1_threshold = density * balance

	# Go 代码是 row-major: index -> x = index % width, y = index / width。
	# Julia 坐标为 1-based；顺序不影响动力学，只影响同 seed 时的初始随机流对应关系。
	for y in 1:gridheight, x in 1:gridwidth
		value = rand(abmrng(model))
		pos = (x, y)

		if value < type1_threshold
			add_agent!(pos, model, 1, "agent1_$(next_type1)")
			next_type1 += 1
		elseif value < density
			add_agent!(pos, model, 2, "agent2_$(next_type2)")
			next_type2 += 1
		end
	end

	return model
end

File: examples/python_mesa/schelling.py
from __future__ import annotations
from typing import List, cast

from mesa import Model
from mesa.datacollection import DataCollector
from mesa.discrete_space import Cell, CellAgent, OrthogonalMooreGrid

DEFAULT_GRID_W = 50
DEFAULT_GRID_H = 50
DEFAULT_DENSITY = 0.8
DEFAULT_BALANCE = 0.5
DEFAULT_SIMILARITY_THRESHOLD = 0.7


class SchellingAgent(CellAgent):
    """One occupied cell in the Schelling grid."""

    cell: "Cell"
    model: "SchellingModel"

    def __init__(self, model: "SchellingModel", cell, group: int) -> None:
        super().__init__(model)
        self.cell = cell
        self.group = group  # 1 or 2; there are no group-0 agents in Mesa

    def is_satisfied(self: SchellingAgent, threshold: float | None = None) -> bool:
        if threshold is None:
            threshold = self.model.similarity_threshold

        same_group = 0
        occupied_neighbors = 0

        for neighbor in self.cell.get_neighborhood(radius=1).agents:
            if cast(SchellingAgent, neighbor).group == self.group:
                same_group += 1
            occupied_neighbors += 1

        if occupied_neighbors == 0:
            return True

        # Same as Go: float64(sameGroup) >= threshold * float64(occupiedNeighbors)
        return same_group >= threshold * occupied_neighbors


class SchellingModel(Model):
    """Mesa rewrite of the provided Go model, preserving the step dynamics."""

    agents: "List[SchellingAgent]"

    def __init__(
        self,
        *,
        width: int = DEFAULT_GRID_W,
        height: int = DEFAULT_GRID_H,
        density: float = DEFAULT_DENSITY,
        balance: float = DEFAULT_BALANCE,
        similarity_threshold: float = DEFAULT_SIMILARITY_THRESHOLD,
        rng=None,
    ) -> None:
        if isinstance(rng, str):
            rng = int(rng) if rng.strip() else None
        super().__init__(rng=rng)

        self.width = width if width > 0 else DEFAULT_GRID_W
        self.height = height if height > 0 else DEFAULT_GRID_H
        self.density = density if 0 <= density <= 1 else DEFAULT_DENSITY
        self.balance = balance if 0 <= balance <= 1 else DEFAULT_BALANCE
        self.similarity_threshold = (
            similarity_threshold
            if 0 <= similarity_threshold <= 1
            else DEFAULT_SIMILARITY_THRESHOLD
        )

        self.grid = OrthogonalMooreGrid(
            (self.width, self.height),
            capacity=1,
            random=self.random,
            torus=False,
        )

        self.initialized = False
        self.last_swapped = 0
        self._populate()
        self.initialized = True

        self.datacollector = DataCollector(
            model_reporters={
                "SatisfiedPct": lambda m: m.satisfied_pct(),
                "SegregationIndex": lambda m: m.segregation_index(),
                "Swapped": "last_swapped",
                "Occupied": lambda m: len(m.agents),
            },
            agent_reporters={
                "group": "group",
                "x": lambda a: a.cell.coordinate[0],
                "y": lambda a: a.cell.coordinate[1],
            },
        )
        self.datacollector.collect(self)

    def _populate(self) -> None:
        """Same per-cell probability rule as the Go Populate method."""
        next_type1 = 0
        next_type2 = 0
        type1_threshold = self.density * self.balance

        # OrthogonalMooreGrid exposes all cells; each may hold at most one agent.
        for cell in self.grid.all_cells:
            value = self.random.random()

            if value < type1_threshold:
                SchellingAgent(self, cell, group=1)
                next_type1 += 1
            elif value < self.density:
                SchellingAgent(self, cell, group=2)
                next_type2 += 1

    def advance(self) -> bool:
        """Global, synchronous-ish selection + paired relocation, matching Go Step()."""

        unsatisfied = [agent for agent in self.agents if not agent.is_satisfied()]
        empty_cells = [cell for cell in self.grid.all_cells if cell.is_empty]

        self.random.shuffle(unsatisfied)
        self.random.shuffle(empty_cells)

        swapped = min(len(unsatisfied), len(empty_cells))
        for agent, new_cell in zip(unsatisfied[:swapped], empty_cells[:swapped]):
            agent.cell = new_cell

        self.last_swapped = swapped
        self.datacollector.collect(self)
        return swapped > 0

    def step(self) -> bool:
        return self.advance()

    def satisfied_pct(self) -> float:
        occupied = len(self.agents)
        if occupied == 0:
            return 0.0
        return sum(1 for a in self.agents if a.is_satisfied()) / occupied

    def segregation_index(self) -> float:
        total_ratio = 0.0
        count = 0

        for agent in self.agents:
            same = 0
            occupied_neighbors = 0

            for neighbor in agent.cell.get_neighborhood(radius=1).agents:
                occupied_neighbors += 1
                if cast(SchellingAgent, neighbor).group == agent.group:
                    same += 1

            if occupied_neighbors > 0:
                total_ratio += same / occupied_neighbors
                count += 1

        return 0.0 if count == 0 else total_ratio / count
